File: 328. Odd Even Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def oddEvenList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head == None:
            return head
        elif head.next == None:
            return head
        elif head.next.next == None:
            return head
        po = head
        pe = head.next
        # more than 2 nodes
        ohead = head.next
        while pe.next != None and pe.next.next != None:
            po.next = pe.next
            po = po.next
            pe.next = po.next
            pe = pe.next
        if pe.next == None:
            po.next = ohead
        else:
            po.next = pe.next
            po.next.next = ohead
            pe.next = None
        return head

File: 328. Odd Even Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 * https://discuss.leetcode.com/topic/34837/simple-c-solution-o-n-time-o-1-space
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if (!head)
            return head;
        ListNode* po = head, *ehead = head -> next, *pe = ehead;
        while(pe && pe -> next){
            po -> next = po -> next -> next;
            pe -> next = pe -> next -> next;
            po = po -> next;
            pe = pe -> next;
        }
        po -> next = ehead;
        return head;
    }
 
};
